// enzyme-adapter-react-helper/build/safeSFC.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _object = require('object.getownpropertydescriptors');

var _object2 = _interopRequireDefault(_object);

var _ifReact = require('./ifReact');

var _ifReact2 = _interopRequireDefault(_ifReact);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function assertFunction(fn) {
  if (typeof fn !== 'function') {
    throw new TypeError('Component must be a function');
  }
  return fn;
}

function copyStatics(source, target) {
  assertFunction(source);

  // eslint-disable-next-line no-param-reassign
  target.displayName = source.displayName || source.name;

  var _gOPDs = (0, _object2['default'])(source),
      oldProto = _gOPDs.prototype,
      descriptors = _objectWithoutProperties(_gOPDs, ['prototype']);

  Object.defineProperties(target, descriptors);

  return target;
}

function nullToNoScript(fn) {
  // eslint-disable-next-line prefer-arrow-callback
  return copyStatics(fn, function () {
    function NullHandler() {
      var element = fn.apply(undefined, arguments);
      return element === null ? _react2['default'].createElement('noscript', null) : element;
    }

    return NullHandler;
  }());
}

var maybeNullWrapper = (0, _ifReact2['default'])('^0.14', nullToNoScript, assertFunction);

function safeSFC(fn) {
  return copyStatics(fn, function (_React$Component) {
    _inherits(SafeSFC, _React$Component);

    function SafeSFC() {
      _classCallCheck(this, SafeSFC);

      return _possibleConstructorReturn(this, (SafeSFC.__proto__ || Object.getPrototypeOf(SafeSFC)).apply(this, arguments));
    }

    _createClass(SafeSFC, [{
      key: 'render',
      value: function () {
        function render() {
          return maybeNullWrapper(fn(this.props, this.context));
        }

        return render;
      }()
    }]);

    return SafeSFC;
  }(_react2['default'].Component));
}

exports['default'] = (0, _ifReact2['default'])('>= 0.14', maybeNullWrapper, safeSFC);
module.exports = exports['default'];
