// react-dfb/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _createFileContent = _interopRequireDefault(require("./create-file-content"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var DownloadButton =
/*#__PURE__*/
function (_React$PureComponent) {
  _inherits(DownloadButton, _React$PureComponent);

  function DownloadButton() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, DownloadButton);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(DownloadButton)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_this), "renderButton", function (_ref) {
      var className = _ref.className,
          style = _ref.style,
          disabled = _ref.disabled,
          onClick = _ref.onClick,
          label = _ref.label;
      var defaultProps = {
        disabled: disabled,
        onClick: onClick
      };
      return _react.default.createElement("div", null, _this.props.children && _react.default.Children.map(_this.props.children, function (child) {
        return _react.default.cloneElement(child, {
          onClick: _this.props.onClick
        });
      }), !_this.props.children && _react.default.createElement("button", _extends({
        type: "button"
      }, defaultProps, {
        onClick: function onClick(event) {
          if (_this.props.disabled) return;

          _this.props.onClick(event);
        }
      }), label), _this.renderDownloadFile());
    });

    _defineProperty(_assertThisInitialized(_this), "renderDownloadFile", function () {
      if (Object.keys(_this.props.downloadData).length === 0) return null;
      var _this$props$downloadD = _this.props.downloadData,
          fileName = _this$props$downloadD.fileName,
          mime = _this$props$downloadD.mime,
          contentBase64 = _this$props$downloadD.contentBase64;
      if (fileName != null && !fileName.length) return null;
      return _react.default.createElement("a", {
        download: fileName,
        href: (0, _createFileContent.default)(contentBase64, mime),
        style: {
          display: 'none'
        },
        ref: function ref(downloadLink) {
          //$FlowFixMe
          downloadLink.click();
        }
      });
    });

    return _this;
  }

  _createClass(DownloadButton, [{
    key: "render",
    value: function render() {
      return _react.default.createElement("div", {
        className: "flex",
        style: {
          display: 'inline-flex'
        }
      }, this.renderButton(this.props));
    }
  }]);

  return DownloadButton;
}(_react.default.PureComponent);

_defineProperty(DownloadButton, "defaultProps", {
  disabled: false,
  downloadData: {
    mime: '',
    fileName: '',
    contentBase64: ''
  },
  label: 'Download!'
});

var _default = DownloadButton;
exports.default = _default;
